### DISCUSSION TITLE
Support for DHARMa residuals (or any alternatives)

### DISCUSSION THREAD

**Discussion by pcinereus** [inlabru-org/inlabru#150] [Ideas]
DHARMa (simulated) residuals are a very useful tool for assisting in the diagnosis of a range of model fitting issues.
It would be great if `inlabru` (and `INLA`) could support these tools.
Following is a MWE illustrating the calculation of DHARMa residuals for `sdmTMB` and my attempt to generate the equivalent in `inlabru`.  It would be great to be able to provide similar diagnostic tools from `inlabru` (and `INLA`).

Thoughts greatly appreciated

Kind regards

murray

```
## load packages
library(tidyverse)
library(INLA)
library(inlabru)
library(sdmTMB)
library(DHARMa)
```

## Fabricated (binomial) data with spatial structure
- COUNT: number of things of a certain type
- TOTAL: total number of things
- LONGITUDE/LATITUDE: spatial coordinates

```
dat <-
structure(list(Longitude = c(147.17854658756832, 148.7228906987358, 
147.9923358716155, 147.68977320452422, 148.38136335386372, 148.05771593831395, 
146.60904030627728, 147.05058379029663, 149.23548015469987, 149.03511715860381, 
148.26705111502417, 147.96960843341475, 147.18342646176382, 146.74532791103942, 
148.42187321971124, 148.7449806096308, 147.76521833404144, 148.63084090562802, 
147.2729338637474, 147.00598024718468), Latitude = c(-19.232046008989595, 
-19.718807977691267, -19.067374258642552, -18.801145988799934, 
-19.597010472932919, -19.774940882708023, -18.370771403649997, 
-19.219430012642867, -19.1980806790754, -19.444344402956276, 
-19.05320271435858, -19.072688420798773, -18.648271075971198, 
-18.467163480297717, -19.56642734909823, -19.207406856169129, 
-19.270339160420978, -19.811894799780102, -18.561375738749891, 
-19.122474978295859), COUNT = c(28L, 31L, 26L, 21L, 13L, 3L, 
18L, 23L, 44L, 24L, 20L, 27L, 21L, 31L, 21L, 33L, 19L, 26L, 30L, 
19L), TOTAL = c(46, 48, 48, 50, 50, 49, 50, 49, 50, 50, 48, 50, 
49, 49, 49, 50, 49, 47, 49, 48)), row.names = c(NA, -20L), class = c("tbl_df", 
"tbl", "data.frame"))
```

## Generate a common mesh

The following is an overly simplistic mesh.  However, it will ensure that it is equivalent between `sdmTMB` and `inlabru`
``` 
mesh.sdmTMB <- make_mesh(dat, c('Longitude','Latitude'), cutoff = 0.1,
                         type = 'cutoff')
mesh <- mesh$sdmTMB   # for INLA
```

## sdmTMB 
Fit spatial (binomial) model
```
mod1 <- sdmTMB(formula = cbind(COUNT, TOTAL-COUNT) ~ 1,
               data = dat,
               mesh = mesh.sdmTMB,
               family = binomial(),
               spatial = "on")
summary(mod1)
```

### DHARMa residuals for sdmTMB
- use `simulate` to generate 250 simulated values (per obs) from the model.  DO NOT include iid effects
- use `predict` to generate predictions (fitted values) marginalizing over all "random effects"
- generate DHARMa simulated residuals (and plot QQ plot and residual plot with quantiles)
- explore spatial autocorrelation (just to illustrate another use of the standardized residuals)
```
sim <- simulate(mod1, nsim = 250, re_form = ~ 0)
pred <- mod1$family$linkinv(predict(mod1,
                                    type = 'link',
                                    re_form = ~0,
                                    re_form_iid = ~0)$est)
mod.resid <- DHARMa::createDHARMa(simulatedResponse = sim,
                                  observedResponse = dat$COUNT,
                                  fittedPredictedResponse = pred)
plot(mod.resid)
DHARMa::testSpatialAutocorrelation(mod.resid,
                                   x = dat$ Longitude,
                                   y = dat$Latitude)
```

## inlabru 
- prepare main data
- fit the model
```
coords <- dat %>%
  dplyr::select(Longitude, Latitude) %>%
  as.matrix()
cmp <- COUNT ~ Intercept +
  spatial.field(main = coords,
                model = mesh.sdmTMB$spde)
mod2 <- bru(cmp, data = dat, family = 'binomial', Ntrials = dat$TOTAL)
summary(mod2)
```

### DHARMA residuals attempt
- use `generate` to attempt to simulate from the model. Unfortunately, I can't work out how to incorporate the uncertainty associated with the hyperparameters (in this case theta) in order to replicate the functionality of the above simulate function (from `sdmTMB`).
- a similar issue would arise for spatio-temporal models
```
sim2 <- generate(mod2, data = dat %>% as.data.frame(),
                 formula = ~ rbinom(20, dat$ TOTAL, plogis(Intercept + spatial.field)),
                 n.samples = 250)
pred2 <- predict(mod2,
                 formula = ~ plogis(Intercept_eval(rep(1, nrow(dat)))),
                 n.samples=250)$mean
mod.resid2 <- DHARMa::createDHARMa(simulatedResponse = sim2,
                                  observedResponse = dat$COUNT,
                                  fittedPredictedResponse = pred2)
plot(mod.resid2)
DHARMa::testSpatialAutocorrelation(mod.resid2,
                                   x = dat$Longitude,
                                   y = dat$Latitude)
```

**Comment by finnlindgren** [comment]
Do you have a good reference/link to what those types of residuals are? It seems very odd to leave out "random effects" from the mean prediction; which components should be left out and which should be included doesn't seem obvious to me, and would be context dependent. Why not the full posterior mean predictions? To automate this there would need to be some way to auto-detect which components to include. For now, leaving it up to the user seems best, as there possible model contructions are so varied.

But it seems to boil down to "simulate posterior samples on the observation level" and "evaluate a (specific type of?) point prediction".
Then your code seems fine (except one `rbinom()` mistake, see example below); there are no additional hyperparameter uncertainty to take into account. For models that _do_ need to explicitly use the hyperparameters for either the generate or predict step, those are currently accessible via their full names (from  `$summary.hyperpar`), but with all special characters replaced by `_` (underscore). I hope to expand that in the future to something like `label_theta`, just like how `label_eval()` works.  This can sometimes be a bit clunky due to the long names:
```r
# The expression should extract needed information from the data,
# which is available in the evaluation environment in `generate()`:
generate(..., data =dat,
              formula = ~ rbinom(nrow(.data.),
                                 size = TOTAL,
                                 prob = plogis(Intercept + spatial.field))
# If I remember the name of the usual Gaussian observation precision correctly:
generate(..., data = dat,
              formula = ~ rnorm(nrow(.data.),
                                mean = Intercept + spatial.field,
                                sd = Precision_for_the_Gaussian_observations^-0.5)
```

These documents indicate that external sampling of this type is the right way. Difficult to wrap for general models, but a short vignett could be used to show how to do it?
https://cran.r-project.org/web/packages/DHARMa/vignettes/DHARMa.html#importing-external-simulations-e.g.-from-bayesian-software-or-unsupported-packages
https://github.com/florianhartig/DHARMa/wiki/Adding-new-R-packages-to-DHARMA

**Reply by pcinereus** [reply]
Thanks so much for your insights @finnlindgren.  My assessment about leaving random effects out of the predictions is based on a comment (see screen grab at the bottom of this post) that appeared in a vignette (<https://pbs-assess.github.io/sdmTMB/articles/residual-checking.html>) produced by the author of the `sdmTMB` package. 
That said, I have just again read through the advice (provided <https://github.com/florianhartig/DHARMa/wiki/Adding-new-R-packages-to-DHARMA> by the author of the `DHARMa` package) on adding DHARMa residual support to other modelling packages. 

To be fair though, some brief testing that I have done when attempting to replicate the simulated residuals associated with `glmmTMB` and `glmer` models does seem to indicate that random effects should be excluded...

> Note for developers:
>
> - I recommend thats signature and output format should conform to the existing implementations in glm and lme4, which is simulate(fittedModel, nsim, ...)
> - ... should ideally allow changing the conditioning of the data creation on all hierarchical levels of the model, which includes being able to condition on the REs, as well as possibly on things like zero-inflation etc.

-----------
![image](https://user-images.githubusercontent.com/16844741/180766605-818adbed-8d83-4d0e-a42d-fe60e79d1b17.png)
-----------

**Reply by finnlindgren** [reply]
They likely have a very specific model structure in mind, for which excluding "the random effects" makes sense; The DHARMa for Bayesian vignette has more to say, and it looks like it's the difference between simulating _new unconditional realisations_ of the random effects, which is very different to _simulating from the posterior distributions of the random effects_. _Excluding_ the random effects is instead setting them to zero, which is different to _both_ of those simulation alternatives! I'd like to see it written in clear maths instead of just referencing code; For some models the random effects have a large effect on the posterior mean of the predictors, so one needs to be careful about it. $E(\exp(x))=\exp(\mu + \sigma^2/2)$, not $\exp(\mu)$, is a common trap...

**Reply by finnlindgren** [reply]
...and that $\sigma^2$ could be either the estimated random effect variance parameter, _or_ the posterior variance (uncertainty) of each individual random effect variable.

Let $\eta_i=z_i u + v_i$ where $z_i$ is a covariate, $u\sim N(0,\infty)$ is the corresponding coefficient, and $v_i \sim N(0,\sigma_v^2)$ are random effects, and let $Y=\{y_i\}$ be observed $\text{Poisson}(\exp(\eta_i))$ values.

Then posterior samples of _new observations_ are given conditionally on posterior samples of the full $\eta_i$ as
$\tilde{y}_i|\eta_i \sim \text{Poisson}(\exp(\eta_i))$
The $\eta_i$ samples include both $u$ and $v_i$ sampling.
For just the posterior predictive expectations, we get (approximatively)
$E(\tilde{y}_i|Y) \approx \exp[z_i E(u|Y) + E(v_i|Y) + Var(z_i u + v_i|Y) / 2 ]$

If the random effect component is excluded, we would get a sampling distribution
$\tilde{\tilde{y}}_i|u,v_i \sim \text{Poisson}(\exp(z_i u))$
and
$E(\tilde{\tilde{y}}_i|Y) \approx \exp[z_i E(u|Y) + Var(z_i u|Y) / 2 ]$
which is completely different.

If we instead sample from the estimated _model_ for the random effect, we get
$\tilde{\tilde{y}}_i|u,\sigma_v^2 \sim \text{Poisson}(\exp(z_i u + w_i))$
where $w_i|\sigma_v^2\sim\text{N}(0,\sigma_v^2)$ with $\sigma_v^2$ sampled from its posterior distribution.
For just the posterior predictive expectations, we get (approximatively)
$E(\tilde{\tilde{y}}_i|Y) \approx \exp[z_i E(u|Y) + Var(z_i u|Y) / 2 +\sigma_v^2/2]$
which again is completely different.

So we would need some more precise explanation of the conceptual mathematics behind the residual definitions to know which version it is that is appropriate to use.  And that may be some fourth version I didn't include above.

**Reply by finnlindgren** [reply]
One approach we've just worked on for point process residuals involved computing the posterior distributions of the "frequentist residuals". For ordinary residuals for the Poisson count case, that leads to a "posterior expected residual" $E[y_i-\exp(\eta_i)|Y]=y_i-E[\exp(\eta_i)|Y]$, which means sample everything from their posterior distributions and compute the predictor expression for each sample, then take the mean. For other types of residuals, it might be different, and there may be other approaches to define Bayesian residual versions.

**Comment by pcinereus** [comment]
I have done a bit of simulating today and think I am getting somewhere.  Will get back to you tomorrow hopefully...
Thanks again..

M
